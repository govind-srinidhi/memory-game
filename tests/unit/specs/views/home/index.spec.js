/**
 * @file It contains unit tests for home page.
 */
import { mount, createLocalVue } from "@vue/test-utils"
import Vuetify from "vuetify"
import Vuex from "vuex"
import VueRouter from "vue-router"
import HomePage from "@/views/home"

let wrapper

const state = {
  userAnswers: Array()
}

const getters = {
  "results/userAnswers": state => state.userAnswers,
}

const mutations = {
  "results/pushUserAnswer": (state, userAnswer) => {
    state.userAnswers.push(userAnswer)
  }
}

const router = new VueRouter()

const mountHomePage = () => {
  const localVue = createLocalVue()

  localVue.use(Vuex)
  localVue.use(Vuetify)
  localVue.use(VueRouter)

  localVue.prototype.$CONSTANTS = {
    LABELS: {
      NUMBER_OF_CARDS: "Number of cards",
      PLAY: "Play"
    },
    GAME_RESULTS: {
      CORRECT: "Correct",
      INCORRECT: "Incorrect"
    }
  }

  const vuetify = new Vuetify()

  const store = new Vuex.Store({
    state,
    getters,
    mutations
  })

  wrapper = mount(HomePage, {
    sync: false,
    localVue,
    store,
    router,
    vuetify
  })
}

const cardsRow = () => wrapper.findComponent({ ref: "row_random_values" })
const playGameButton = () => wrapper.findComponent({ ref: "button_play_game" })

describe("home page", () => {
  beforeEach(() => {
    VueRouter.prototype.replace = jest.fn()
    mountHomePage()
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it("renders the home page", () => {
    expect(wrapper).toBeTruthy()
  })

  it("renders elements correctly", async () => {
    expect(wrapper.findAll("#select_cards").at(0).exists()).toBeTruthy()
    expect(playGameButton().exists()).toBeTruthy()
  })

  it("generates random values same as the number of cards selected in dropdown", async () => {
    const numberOfCards = 4
    const input = wrapper.findAll("#select_cards").at(0)
    input.element.value = numberOfCards
    input.trigger("input")
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.randomValues.length).toBe(numberOfCards)
    expect(wrapper.vm.isGameStarted).toEqual(false)
  })

  it("shows random values generated on each of the card until user clicks play button", async () => {
    const numberOfCards = 4
    const input = wrapper.findAll("#select_cards").at(0)
    input.element.value = numberOfCards
    input.trigger("input")
    await wrapper.vm.$nextTick()
    const randomValues = wrapper.vm.randomValues
    const cardColumns = cardsRow().findAll(".col")
    for (var i = 0; i < cardColumns.length; i++) {
      expect(parseInt(cardColumns.at(i).find("span").text())).toEqual(randomValues[i].value)
    }
  })

  it("the cards should become blank, once user clicks on play button", async () => {
    const numberOfCards = 4
    const input = wrapper.findAll("#select_cards").at(0)
    input.element.value = numberOfCards
    input.trigger("input")
    await wrapper.vm.$nextTick()
    playGameButton().trigger("click")
    await wrapper.vm.$nextTick()
    const randomValues = wrapper.vm.randomValues
    const cardColumns = cardsRow().findAll(".col")
    expect(cardColumns.length).toEqual(randomValues.length)
    for (var i = 0; i < cardColumns.length; i++) {
      expect(cardColumns.at(i).find("span").exists()).toBe(false)
    }
  })

  it("as user clicks on each of the cards, need to validate, if the order followed is correct/incorrect and navigate to results", async () => {
    const numberOfCards = 4
    const input = wrapper.findAll("#select_cards").at(0)
    input.element.value = numberOfCards
    input.trigger("input")
    await wrapper.vm.$nextTick()
    playGameButton().trigger("click")
    await wrapper.vm.$nextTick()
    const sortedRandomValues = wrapper.vm.sortedRandomValues
    const cardColumns = cardsRow().findAll(".col")
    for (var i = 0; i < cardColumns.length; i++) {
      const card = cardColumns.at(i).find(".v-card")
      card.trigger("click")
      await wrapper.vm.$nextTick()
      if (sortedRandomValues[i] == cardColumns.at(i).find("span").text()) {
        expect(card.find(".validation-message").classes()).toContain("success--text")
      } else {
        expect(card.find(".validation-message").classes()).toContain("error--text")
      }
    }
    expect(router.replace).toBeCalledWith("results")
  })

  it("matches the default snapshot", async () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})