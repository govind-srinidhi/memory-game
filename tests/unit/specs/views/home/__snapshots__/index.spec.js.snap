// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home page matches the default snapshot 1`] = `
<div class="container pa-3 ml-6">
  <div class="row">
    <div class="col-md-3 col">
      <div class="v-input theme--light v-text-field v-select">
        <div class="v-input__control">
          <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-80" class="v-input__slot">
            <div class="v-select__slot"><label for="select_cards" class="v-label theme--light" style="left: 0px; position: absolute;">Number of cards</label>
              <div class="v-select__selections"><input id="select_cards" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
              </div><input type="hidden" value="-1">
            </div>
            <div class="v-menu">
              <!---->
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="col-md-3 col"><button type="button" disabled="disabled" class="mt-4 v-btn v-btn--disabled v-btn--has-bg theme--light v-size--default"><span class="v-btn__content">Play</span></button></div>
  </div>
  <div class="row"></div>
</div>
`;
